multigeiger/ble.cpp

```diff
@@ -0,0 +1,138 @@
+BLEDescriptor bleDescriptorHRPOS(BLE_DESCR_UUID);
+
+static bool ble_enabled = false;
+static bool device_connected = false;
+
+uint8_t flags_HRS = 0b00001001; // bit 0 --> 1 = HR (cpm) as UINT16, bit 3 --> add UINT16 Energy Expended (rolling update counter)
+uint8_t txBuffer_HRM[5]; // transmit buffer, byte[0] = flags, [1, 2] = cpm, [3, 4] = rolling update counter
+uint8_t txBuffer_HRPOS[1] = {TUBE_TYPE};
+unsigned int status_HRCP = 0;
+unsigned int cpm_update_counter = 0;
+unsigned int cpm_value = 0;
+
+bool is_ble_connected() {
+  return ble_enabled && device_connected;
+}
+
+class MyServerCallbacks: public BLEServerCallbacks {
+  void onConnect(BLEServer *pServer, esp_ble_gatts_cb_param_t *param) {
+    char remoteAddress[18];
+    sprintf(
+      remoteAddress,
+      "%02X:%02X:%02X:%02X:%02X:%02X",
+      param->connect.remote_bda[0],
+      param->connect.remote_bda[1],
+      param->connect.remote_bda[2],
+      param->connect.remote_bda[3],
+      param->connect.remote_bda[4],
+      param->connect.remote_bda[5]
+    );
+    log(INFO, "BLE device connected, remote MAC: %s", remoteAddress);
+    device_connected = true;
+  };
+
+  void onDisconnect(BLEServer *pServer) {
+    device_connected = false;
+    log(INFO, "BLE device disconnected");
+  }
+};
+
+// Callback allowing for Control Point Characteristic to reset "energy expenditure" (packet counter)
+class MyCharacteristicCallbacks: public BLECharacteristicCallbacks {
+  void onWrite(BLECharacteristic *pCharacteristic) {
+    std::string rxValue = pCharacteristic->getValue();
+    status_HRCP = (rxValue.length() > 0) ? rxValue[0] : 0;
+  }
+};
+
+void update_bledata(unsigned int cpm) {
+  if (!ble_enabled)
+    return;
+  cpm_update_counter++;
+  cpm_update_counter = cpm_update_counter & 0xFFFF;
+  if (status_HRCP > 0) {
+    log(DEBUG, "HR Control Point Value received: %d, resetting packet counter", status_HRCP);
+    cpm_update_counter = 0;
+    status_HRCP = 0;
+  }
+  cpm_value = cpm;
+  txBuffer_HRM[0] = flags_HRS;
+  txBuffer_HRM[1] = cpm_value & 0xFF;
+  txBuffer_HRM[2] = (cpm_value >> 8) & 0xFF;
+  txBuffer_HRM[3] = cpm_update_counter & 0xFF;
+  txBuffer_HRM[4] = (cpm_update_counter >> 8) & 0xFF;
+  bleCharHRM.setValue(txBuffer_HRM, 5);
+  bleCharHRM.notify();
+}
+
+void setup_ble(char *device_name, bool ble_on) {
+  if (!ble_on) {
+    set_status(STATUS_BT, ST_BT_OFF);
+    return;
+  }
+  ble_enabled = true;
+
+  set_status(STATUS_BT, ST_BT_INIT);
+  BLEDevice::init(device_name);
+
+  BLEServer *bleServer = BLEDevice::createServer();
+  bleServer->setCallbacks(new MyServerCallbacks());
+
+  BLEService *bleService = bleServer->createService(BLE_SERVICE_HEART_RATE);
+
+  bleService->addCharacteristic(&bleCharHRM);
+  bleDescriptorHRM.setValue("Radiation rate CPM");
+  bleCharHRM.addDescriptor(&bleDescriptorHRM);
+  bleCharHRM.addDescriptor(new BLE2902());  // required for internal notification management
```

Comment:
"internal" refers to what?

Reply:
Service internal. Rephrased for better understanding.